src/Liquid.OnAzure/Storages/AzureBlob.cs

```
@@ -15,62 +15,124 @@ namespace Liquid.OnAzure
+            }
+
+            Initialize(configuration);
         }
 
+        /// <summary>
+        /// Initializes the object. 
+        /// </summary>
+        /// <remarks>
+        /// This method obtains the configuration using <see cref="LightConfigurator"/> directly.
+        /// </remarks>
         public void Initialize()
         {
-            //Get the configuration on appsetting. But in this case the features can be accessed outside from the repository
-            this.mediaStorageConfiguration = LightConfigurator.Config<MediaStorageConfiguration>("MediaStorage");
-            this.Conection = mediaStorageConfiguration.ConnectionString;
-            this.Permission = mediaStorageConfiguration.Permission;
-            this.Container = mediaStorageConfiguration.Container;
+            // TODO: mark this as obsolete and remove it.
+
+            // Get the configuration on appsetting. But in this case the features can be accessed outside from the repository
+            var configuration = LightConfigurator.Config<MediaStorageConfiguration>("MediaStorage");
+            Initialize(configuration);
+        }
+
+        /// <summary>
+        /// Initializes the class based on the provided configuration.
+        /// </summary>
+        /// <param name="configuration">The configuiration for this class.</param>
+        private void Initialize(MediaStorageConfiguration configuration)
+        {
+            _connection = configuration.ConnectionString;
+            _container = configuration.Container;
+            _permission = configuration.Permission;
+
+            SetContainerReference(_container);
 
             // If the MS has the configuration outside from the repository will be used this context and not inside
-            if (WorkBench.Repository != null)
-            {
-                WorkBench.Repository.SetMediaStorage(this);
-            }
+            WorkBench.Repository.SetMediaStorage(this);
```

I see that you removed the null check, which seems completely fine for me since I think there will never be a scenario where the global repository is null, but is it true though? I'm just asking because I'm not 100% sure about that.

---

I no longer remember my rationale for removing the null check, but what I can see is that the code fails silently. In fact, the whole behavior is very hidden from the user:

> If there's a repository set by the time we initialize `AzureBlob`, then set it as the official MediaStorage, otherwise, skip.

I think that we should either fail - it's null, there's no repository, that's an issue - or set the class as an repository. 

In any case, again, I think that this is better solved by leveraging DI.

---

All that said, what do you think? Do I bring back the null checking as it was, or create a null check that fails with a decent error message?

---

@bruno-brant I took a time to track this behavior and I think we have a bigger problem.

As we can see (and as you've mentioned), this class will be set as the official MediaStorage if there's a repository set up. Workbench services (Repository, Cache, Telemetry, ...) are defined by using the Workbench methods such as `UseRepository` or `UseMessageBus`. Those services are then initialized by calling the `GetRegisteredService` method

https://github.com/Avanade/Liquid-Application-Framework/blob/ac834f46809cd445dda8783ff616ea4ae7af6b33/src/Liquid.Base/Workbench.cs#L206-L212

which calls the `InitializeServices` method.

https://github.com/Avanade/Liquid-Application-Framework/blob/ac834f46809cd445dda8783ff616ea4ae7af6b33/src/Liquid.Base/Workbench.cs#L263-L270

As you can see, it just iterates over the  `_singletonCache`. We can spot the issue here. Well, if I register the Repository **after** the MediaStorage, that check would never evaluate to true and your code would certainly throw a `NullReferenceException`.

What should we do in this case (since this service clearly depends on the Repository instance)? One could argue that Liquid has an aggressive opinionated programming model that also force us to always register the repository before the mediastorage, but that would be insane.

---

Yeah, that the expanded explanation of what I meant by "If there's a repository set *by the time* we initialize AzureBlob, then set it as the official MediaStorage, otherwise, skip.". (check the emphasis that I added). 

That's why I prefer to fail if Repository is null, to avoid this kind of surprising behavior. However, this may lead to a breaking change - maybe the user decided to use a media storage explicity (accessing it from `WorkBench.MediaStorage`) directly from his/her code, but made sure that the MediaStorage used by Repository isn't changed. 

....

All that considered, I'd play the `alpha` card here: we can assume (kinda safely) that IMediaStorage is supposed to always replace the default implementation when set, therefore, should always come _after_ Repository.

That would still lead to bad errors that our user might have problem with; my suggestion is to add a exception that informs the user that the repository must be set *before* the MediaStorage or the code won't work;

While this is bad, we're slowly creeping towards DI (I mean to have that done by march or april) which would solve this (no more globally setting the MediaStorage from the media storage initialization).

---

Hmm, I've just realized that there's a better solution: we can remove the SetMediaStorage completely and make Repository access MediaStorage from the Workbench singleton instance.

---

After all this discussing, I decided to revert back to the original code (with the null checking). The reason is simply because, to fix this, I need to deal with CosmosDb, which needs a lot of unit testing before I could be comfortable making any changes.

So I updated the discussion I had on this issue (#73) to describe the actual bug that we detected, and it will be dealt with at a later time - after we create the necessary unit tests.